plugins/main/public/components/common/hooks/data_grid/data_grid_service.ts

```diff
@@ -0,0 +1,163 @@
+    const DEFAULT_MAX_SIZE_PER_CALL = 1000;
+    const { indexPattern, filters = [], query, sorting, fields, pagination } = params;
+    // when the pageSize is greater than the default max size per call (10000)
+    // then we need to paginate the search
+    const mustPaginateSearch = pagination?.pageSize && pagination?.pageSize > DEFAULT_MAX_SIZE_PER_CALL;
+    const pageSize = mustPaginateSearch ? DEFAULT_MAX_SIZE_PER_CALL : pagination?.pageSize;
+    const totalHits = pagination?.pageSize || DEFAULT_MAX_SIZE_PER_CALL; 
+    let pageIndex = params.pagination?.pageIndex || 0;
+    let hitsCount = 0;
+    let allHits = [];
+    let searchResults;
+    if (mustPaginateSearch) {
+        // paginate the search
+        while (hitsCount < totalHits &&  hitsCount < MAX_ENTRIES_PER_QUERY) {
+            const searchParams = {
+                indexPattern,
+                filters,
+                query,
+                pagination: {
+                    pageIndex,
+                    pageSize,
+                },
+                sorting,
+                fields,
+            };
+            searchResults = await search(searchParams);
+            allHits = allHits.concat(searchResults.hits.hits);
+            hitsCount = allHits.length;
+            pageIndex++;
+        }
+    } else {
+        searchResults = await search(params);
+        allHits = searchResults.hits.hits;
+    }
+    
+    const resultsFields = fields;
+    const data = allHits.map((hit) => { 
+        // check if the field type is a date
+        const dateFields = indexPattern.fields.getByType('date');
+        const dateFieldsNames = dateFields.map((field) => field.name);
+        const flattenHit = indexPattern.flattenHit(hit);
+        // replace the date fields with the formatted date
+        dateFieldsNames.forEach((field) => {
+            if (flattenHit[field]) {
+                flattenHit[field] = beautifyDate(flattenHit[field]);
+            }
+        });
+        return flattenHit;
+    });
+
+    if (!resultsFields || resultsFields.length === 0){
+        return;
+    }
+
+    if (!data || data.length === 0)
+        return;
+
+    const parsedData = data.map((row) => {
+        const parsedRow = resultsFields?.map((field) => {
+            const value = row[field];
+            if (value === undefined || value === null) {
+                return '';
+            }
+            if (typeof value === 'object') {
+                return JSON.stringify(value);
+            }
+            return `"${value}"`;
+        });
+        return parsedRow?.join(',');
+    }).join('\n');
+
+    // create a csv file using blob
+    const blobData = new Blob(
+        [
+            `${resultsFields?.join(',')}\n${parsedData}`
+        ],
+        { type: 'text/csv' }
+    );
+
+    if (blobData) {
+        // @ts-ignore
+        FileSaver?.saveAs(blobData, `vulnerabilities_inventory-${new Date().toISOString()}.csv`);
```

Comment:
The text refers to vulnerabilities but this will be used in other modules.

Reply:
Thanks @chantal-kelm